slatedb/src/db_state.rs

```diff
@@ -23,50 +23,99 @@ use SsTableId::{Compacted, Wal};
 pub(crate) struct SsTableHandle {
     pub id: SsTableId,
     pub info: SsTableInfo,
+    pub visible_range: Option<BytesRange>,
+    effective_range: BytesRange,
 }
 
 impl Debug for SsTableHandle {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        f.write_fmt(format_args!("SsTableHandle({:?})", self.id))
+        f.write_fmt(format_args!(
+            "SsTableHandle({:?}, {:?})",
+            self.id, self.visible_range
+        ))
     }
 }
 
 impl SsTableHandle {
     pub(crate) fn new(id: SsTableId, info: SsTableInfo) -> Self {
-        SsTableHandle { id, info }
+        let effective_range = match info.first_key.clone() {
+            Some(physical_first_key) => BytesRange::new(Included(physical_first_key), Unbounded),
+            None => {
+                // Empty range.
+                BytesRange::new(
+                    Excluded(Bytes::copy_from_slice(&[0_u8])),
+                    Excluded(Bytes::copy_from_slice(&[0_u8])),
+                )
+            }
+        };
+
+        SsTableHandle {
+            id,
+            info,
+            visible_range: None,
+            effective_range,
+        }
+    }
+
+    pub(crate) fn new_compacted(
+        id: SsTableId,
+        info: SsTableInfo,
+        visible_range: Option<BytesRange>,
+    ) -> Self {
+        let mut effective_range = match info.first_key.clone() {
+            Some(physical_first_key) => {
+                BytesRange::new(Included(physical_first_key.clone()), Unbounded)
+            }
+            None => {
+                unreachable!("SST always has a first key.")
```

Comment:
🤔 Is this true? I wonder what the compactor does if the only `RowEntry`s in the SST are deletes....

Reply:
here is a test case; what's little weird is that we'll end up with "empty sorted run" entry, but that's most likely harmless